service/history/historyReplicator.go

```diff
@@ -78,21 +78,26 @@ func (r *historyReplicator) ApplyEvents(request *h.ReplicateEventsRequest) (retE
 
 	execution := *request.WorkflowExecution
 
-	var context *workflowExecutionContext
+	context, release, err := r.historyCache.getOrCreateWorkflowExecution(domainID, execution)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
 	var msBuilder *mutableStateBuilder
 	firstEvent := request.History.Events[0]
 	switch firstEvent.GetEventType() {
 	case shared.EventTypeWorkflowExecutionStarted:
+		msBuilder, err = context.loadWorkflowExecution()
+		if err == nil {
```

Comment:
maybe need a type assertion here, i.e. entity not exist

Reply:
Good catch.  Let me add that.